hw_5.py

```diff
@@ -0,0 +1,167 @@
+             regex=True)
+alignment = read_bed6('data/alignment.bed')
+
+rna_23s = rrna_annotation[rrna_annotation['attributes'] == '23S']['chromosome'].value_counts().rename('23S')
+rna_16s = rrna_annotation[rrna_annotation['attributes'] == '16S']['chromosome'].value_counts().rename('16S')
+rna_5s = rrna_annotation[rrna_annotation['attributes'] == '5S']['chromosome'].value_counts().rename('5S')
+
+rna_count_table = pd.concat([rna_5s, rna_16s, rna_23s], axis = 1) \
+    .sort_index() \
+    .reset_index(level=0) \
+    .rename({'index' : 'Sequence'}, axis = 'columns')
+
+plt.figure(0)
+rna_count_table.plot(x='Sequence', y=['16S', '23S', '5S'], kind='bar', figsize=(20,10))
+plt.ylabel('Count')
+plt.legend(title='RNA type')
+plt.show()
+
+rrna_annotation['annotation_interval'] = rrna_annotation.apply(lambda row: (row.start, row.end), axis = 1)
+alignment['alignment_interval'] = alignment.apply(lambda row: (row.start - 1, row.end + 1), axis = 1)
+merged = rrna_annotation.merge(alignment, on='chromosome')
+merged = merged[merged.apply(lambda row : 
+                                 row.alignment_interval[0] >= row.annotation_interval[0] and 
+                                 row.alignment_interval[1] <= row.annotation_interval[1],
+                             axis = 1)]
+merged.drop(['annotation_interval', 'alignment_interval'], axis=1, inplace=True)
+merged
+
+"""## Задание 2"""
+
+from matplotlib.ticker import AutoMinorLocator
+from matplotlib import rcParams
+import heapq
+
+diffexpr_data = pd.read_csv('data/diffexpr_data.tsv.gz', sep='\t')
+
+def regulation(table: pd.DataFrame) -> str:
+    log_fc, p_val = table
+    if log_fc < 0 and p_val < 0.05:
+        return 'Significantly down-regulated'
+    if log_fc < 0 and p_val > 0.05:
+        return 'Non-significantly down-regulated'
+    if log_fc > 0 and p_val < 0.05:
+        return 'Significantly up-regulated'
+    return 'Non-significantly up-regulated'
+
+diffexpr_data['regulation'] = diffexpr_data[['logFC', 'pval_corr']].apply(regulation, axis=1)
+
+colors_map = {
+    'Significantly down-regulated': '#1F77B4', 
+    'Significantly up-regulated': '#FF7F0E', 
+    'Non-significantly down-regulated': '#2CA02C', 
+    'Non-significantly up-regulated': '#D62728'
+}
+
+bound = max(abs(diffexpr_data['logFC'])) + 1
+plot_x_limits = [-bound, bound]
+
+def significant_points(df: pd.DataFrame):
+    up_regulated = df.loc[(df['regulation'] == 'Significantly up-regulated') & (df['pval_corr'] < 0.05)]
+    down_regulated = df.loc[(df['regulation'] == 'Significantly down-regulated') & (df['pval_corr'] < 0.05)]
+    reg = heapq.nlargest(2, up_regulated['logFC']) + [-x for x in heapq.nlargest(2, abs(down_regulated['logFC']))]
+    return map(lambda reg_item: (
+        reg_item, 
+        df.loc[df['logFC'] == float(reg_item), 'Sample'].iloc[0],
+        df.loc[df['logFC'] == float(reg_item), 'log_pval'].iloc[0]
+    ), reg)
+
+significant = significant_points(diffexpr_data)
```

Comment:
Есть чуть более простые стопобы найти топ-2 генов)
```suggestion
top_upregulated = diffexpr_data.query("pval_corr < 0.05").nlargest(2, "logFC")
top_downregulated = diffexpr_data.query("pval_corr < 0.05").nsmallest(2, "logFC")
```